internal/service/searchdeployment/state_transition_search_deployment_test.go

```diff
@@ -3,155 +3,126 @@ package searchdeployment_test
-				{
-					DeploymentResp: responseWithState("UPDATING"),
-				},
-				{
-					DeploymentResp: responseWithState("IDLE"),
-				},
+			mockResponses: []response{
+				{state: &updating},
+				{state: &idle},
 			},
-			expectedResult: responseWithState("IDLE"),
-			expectedError:  false,
+			expectedState: &idle,
+			expectedError: false,
 		},
 		{
 			name: "Successful transition to IDLE with 503 error in between",
-			mockResponses: []SearchDeploymentResponse{
-				{
-					DeploymentResp: responseWithState("UPDATING"),
-				},
-				{
-					DeploymentResp: nil,
-					HTTPResponse:   &http.Response{StatusCode: 503},
-					Err:            errors.New("Service Unavailable"),
-				},
-				{
-					DeploymentResp: responseWithState("IDLE"),
-				},
+			mockResponses: []response{
+				{state: &updating},
+				{statusCode: sc503, err: errors.New("Service Unavailable")},
+				{state: &idle},
 			},
-			expectedResult: responseWithState("IDLE"),
-			expectedError:  false,
+			expectedState: &idle,
+			expectedError: false,
 		},
 		{
 			name: "Error when transitioning to an unknown state",
-			mockResponses: []SearchDeploymentResponse{
-				{
-					DeploymentResp: responseWithState("UPDATING"),
-				},
-				{
-					DeploymentResp: responseWithState(""),
-				},
+			mockResponses: []response{
+				{state: &updating},
+				{state: &unknown},
 			},
-			expectedResult: nil,
-			expectedError:  true,
+			expectedState: nil,
+			expectedError: true,
 		},
 		{
 			name: "Error when API responds with error",
-			mockResponses: []SearchDeploymentResponse{
-				{
-					DeploymentResp: nil,
-					HTTPResponse:   &http.Response{StatusCode: 500},
-					Err:            errors.New("Internal server error"),
-				},
+			mockResponses: []response{
+				{statusCode: sc500, err: errors.New("Internal server error")},
 			},
-			expectedResult: nil,
-			expectedError:  true,
+			expectedState: nil,
+			expectedError: true,
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			mockService := MockSearchDeploymentService{
-				MockResponses: tc.mockResponses,
-			}
-
-			resp, err := searchdeployment.WaitSearchNodeStateTransition(context.Background(), dummyProjectID, "Cluster0", &mockService, testTimeoutConfig)
-
-			if (err != nil) != tc.expectedError {
-				t.Errorf("Case %s: Received unexpected error: %v", tc.name, err)
-			}
-
-			if !reflect.DeepEqual(tc.expectedResult, resp) {
-				t.Errorf("Case %s: Response did not match expected output", tc.name)
+			svc := mocksvc.NewDeploymentService(t)
+			ctx := context.Background()
+			for _, resp := range tc.mockResponses {
+				svc.On("GetAtlasSearchDeployment", ctx, dummyProjectID, clusterName).Return(resp.get()...).Once()
 			}
+			resp, err := searchdeployment.WaitSearchNodeStateTransition(ctx, dummyProjectID, "Cluster0", svc, testTimeoutConfig)
+			assert.Equal(t, tc.expectedError, err != nil)
+			assert.Equal(t, responseWithState(tc.expectedState), resp)
+			svc.AssertExpectations(t)
```

Comment:
check mock expectations

Reply:
Just want to point out that "AssertExpectations asserts that everything specified with On and Return was in fact called as expected.  Calls may have occurred in any order." ([link](https://github.com/stretchr/testify/blob/db8608ed63f5bd9abdd03f669f5bb80569c114b6/mock/mock.go#L594C4-L595C70)) In some tests we might want to avoid this call and instead use other methods such as `AssertNumberOfCalls`, `AssertNotCalled`, etc.

Reply:
thanks for the clarification